Enable Zeta edit predictions with custom URL without authentication

DominicBurkart opened the pull request:
Enables using Zeta edit predictions with a custom `ZED_PREDICT_EDITS_URL` without requiring authentication to Zed servers. This is useful for:

- Development and testing workflows
- Self-hosted Zeta instances
- Custom AI model endpoints

Prior context on this usage of `ZED_PREDICT_EDITS_URL`: https://github.com/zed-industries/zed/pull/30418

Release Notes:
- Improved self-hosted zeta UX. Users no longer have to log into Zed to use custom or self-hosted zeta backends.

nicky-isaacs-awoo commented on crates/zed/src/zed/edit_prediction_registry.rs:
```diff
@@ -238,7 +238,9 @@ fn assign_edit_prediction_provider(
                     editor.set_edit_prediction_provider(Some(provider), window, cx);
-                } else if user_store.read(cx).current_user().is_some() {
+                } else if user_store.read(cx).current_user().is_some()
+                    || std::env::var("ZED_PREDICT_EDITS_URL").is_ok()
```
can `ZED_PREDICT_EDITS_URL` be pulled into a constant?

nicky-isaacs-awoo reviewed (commented):
LGTM but very much not an expert

DominicBurkart committed:
test(zeta): Add authentication bypass tests for custom URLs

Add tests to verify authentication behavior with custom URLs:
- test_unauthenticated_without_custom_url_blocks_prediction: verifies
  that without authentication and without a custom URL, prediction fails
- test_unauthenticated_with_custom_url_allows_prediction: verifies that
  with a custom URL set via ZED_PREDICT_EDITS_URL, prediction should
  proceed even without authentication

These tests establish the expected behavior: when using custom URLs,
authentication should be optional (use if available, proceed without).

DominicBurkart committed:
feat(zeta): Allow custom URLs without authentication

When ZED_PREDICT_EDITS_URL is set (custom URL), authentication is now
optional - the prediction will use authentication if available, and
proceed without it if not available.

When no custom URL is set (standard Zed infrastructure), authentication
remains required to maintain the existing behavior.

Changes:
- Token acquisition uses .ok() for custom URLs (converts error to None)
- Authorization header only added when token is present
- Token refresh only attempted when we had a token to begin with

DominicBurkart committed:
refactor(zeta): Make auth tests CI-compatible

Add test-only infrastructure to allow testing custom URL auth behavior
without requiring environment variable manipulation:

- Add TEST_FORCE_CUSTOM_URL_MODE AtomicBool with #[cfg(test)]
- Add has_custom_url() helper that checks test override first
- Combine auth tests into single test_authentication_behavior function
  for sequential execution (avoids concurrency issues)
- Remove #[ignore] from custom URL test - now runs in CI

The test-only code has zero impact on production builds.

DominicBurkart committed:
style: Apply cargo fmt to edit_prediction_tests

agu-z committed:
Merge branch 'main' into dominic.burkart/zeta-works-while-signed-out-if-ZED_PREDICT_EDITS_URL-set

agu-z committed:
Refactor custom url/auth handling

agu-z committed:
Fix var name

agu-z commented:
@DominicBurkart Thank you for your contribution! Merging after some tweaks

agu-z committed:
Fix test

SuperKenVery commented:
@pjv You may want to use this [proxy](https://github.com/g0t4/zed-zeta-server) to wrap ollama, then use `PREDICT_EDITS_URL` to point zed to it.
